### PR TITLE
HTTP tests and Elasticurl.NET

### PR DESCRIPTION
**Pull request by justinboswell**
Dependent on: https://github.com/awslabs/aws-c-http/pull/40

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tools/Elasticurl/Elasticurl.cs [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+        {
+            Logger.EnableLogging(ctx.LogLevel, ctx.TraceFile);
+        }
+
+        static void InitOutput()
+        {
+            if (ctx.OutputFilename != null)
+            {
+                try {
+                    ctx.OutputStream = File.OpenWrite(ctx.OutputFilename);
```
Comment (JonathanHenson) [inline tools/Elasticurl/Elasticurl.cs]: do you need the binary flag on this one?

File: tools/Elasticurl/Elasticurl.cs [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+        static void InitLogging()
+        {
+            Logger.EnableLogging(ctx.LogLevel, ctx.TraceFile);
+        }
+
+        static void InitOutput()
+        {
+            if (ctx.OutputFilename != null)
+            {
+                try {
```
Comment (JonathanHenson) [inline tools/Elasticurl/Elasticurl.cs]: style consistency

File: tools/Elasticurl/Elasticurl.cs [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+
+        static TlsConnectionOptions InitTls()
+        {
+            TlsConnectionOptions tlsConnectionOptions = null;
+            if (ctx.Uri.Scheme == Uri.UriSchemeHttps || (ctx.Uri.Port != 80 && ctx.Uri.Port != 8080))
+            {
+                TlsContextOptions tlsOptions = null;
+                if (ctx.Certificate != null && ctx.PrivateKey != null)
+                {
+                    try {
```
Comment (JonathanHenson) [inline tools/Elasticurl/Elasticurl.cs]: style

File: aws-crt-http/HttpConnection.cs [hunk 4]
```diff
@@ -138,22 +147,24 @@ public HttpClientStream(HttpClientConnection connection, HttpRequestOptions opti
                 throw new ArgumentNullException("OnStreamComplete");
             
             // Wrap the native callbacks to bind this stream to them as the first argument
-            OnStreamOutgoingBodyNative onStreamOutgoingBody = (out byte[] buffer, out UInt64 size) =>
+            OnStreamOutgoingBodyNative onStreamOutgoingBody = (byte[] buffer, UInt64 size, out UInt64 bytesWritten) =>
             {
-                options.OnStreamOutgoingBody(this, out buffer);
-                size = (UInt32)buffer.Length;
+                return (int)options.OnStreamOutgoingBody(this, buffer, out bytesWritten);
```
[the remark this answers is not in the document]
Comment (justinboswell) [reply aws-crt-http/HttpConnection.cs]: That's my next job

File: tools/Elasticurl/Elasticurl.cs (hunk 1 above)
Comment (justinboswell) [reply tools/Elasticurl/Elasticurl.cs]: FileStream is binary

File: native/src/http_client.c [hunk 5]
```diff
@@ -174,10 +183,11 @@ static void s_stream_on_incoming_body(
     size_t *out_window_update_size,
     void *user_data) {
     (void)s;
-    (void)out_window_update_size;
     struct aws_dotnet_http_stream *stream = user_data;
     if (stream->on_incoming_body) {
-        stream->on_incoming_body(data->ptr, data->len);
+        uint64_t window_update_size = *out_window_update_size;
+        stream->on_incoming_body(data->ptr, data->len, &window_update_size);
+        *out_window_update_size = window_update_size;
```
Comment (graebm) [inline native/src/http_client.c]: is this necessary for super-safety?
`*out_window_update_size = window_update_size < SIZE_MAX ? window_update_size : SIZE_MAX;`

Comment (justinboswell) [reply native/src/http_client.c]: Always be safe, kids.

**Pull request merged**